plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal.hs

```diff
@@ -0,0 +1,251 @@
+  | LogShake Shake.Log
+  | LogDocOpened Uri
+  | LogDocModified Uri
+  | LogDocSaved Uri
+  | LogDocClosed Uri
+  deriving Show
+
+instance Pretty Log where
+  pretty = \case
+    LogShake log' -> pretty log'
+    LogModificationTime nfp modTime  ->
+      "Modified:" <+> pretty (fromNormalizedFilePath nfp) <+> pretty (show modTime)
+    LogDiagnostics nfp diags ->
+      "Diagnostics for" <+> pretty (fromNormalizedFilePath nfp) <> ":" <+> pretty (show diags)
+    LogDocOpened uri ->
+      "Opened text document:" <+> pretty (getUri uri)
+    LogDocModified uri ->
+      "Modified text document:" <+> pretty (getUri uri)
+    LogDocSaved uri ->
+      "Saved text document:" <+> pretty (getUri uri)
+    LogDocClosed uri ->
+      "Closed text document:" <+> pretty (getUri uri)
+
+descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
+descriptor recorder plId = (defaultCabalPluginDescriptor plId)
+  { pluginRules = cabalRules recorder
+  , pluginHandlers = mkPluginHandler STextDocumentCodeAction licenseSuggestCodeAction
+  , pluginNotificationHandlers = mconcat
+  [ mkPluginNotificationHandler LSP.STextDocumentDidOpen $
+      \ide vfs _ (DidOpenTextDocumentParams TextDocumentItem{_uri,_version}) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocOpened _uri
+        addFileOfInterest ide file Modified{firstOpen=True}
+        restartCabalShakeSession ide vfs file "(opened)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidChange $
+      \ide vfs _ (DidChangeTextDocumentParams VersionedTextDocumentIdentifier{_uri} _) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocModified _uri
+        addFileOfInterest ide file Modified{firstOpen=False}
+        restartCabalShakeSession ide vfs file "(changed)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidSave $
+      \ide vfs _ (DidSaveTextDocumentParams TextDocumentIdentifier{_uri} _) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocSaved _uri
+        addFileOfInterest ide file OnDisk
+        restartCabalShakeSession ide vfs file "(saved)"
+
+  , mkPluginNotificationHandler LSP.STextDocumentDidClose $
+      \ide vfs _ (DidCloseTextDocumentParams TextDocumentIdentifier{_uri}) -> liftIO $ do
+      whenUriFile _uri $ \file -> do
+        log' Debug $ LogDocClosed _uri
+        deleteFileOfInterest ide file
+        restartCabalShakeSession ide vfs file "(closed)"
+  ]
+  }
+  where
+    log' = logWith recorder
+
+    whenUriFile :: Uri -> (NormalizedFilePath -> IO ()) -> IO ()
+    whenUriFile uri act = whenJust (LSP.uriToFilePath uri) $ act . toNormalizedFilePath'
+
+-- | Helper function to restart the shake session, specifically for modifying .cabal files.
+-- No special logic, just group up a bunch of functions you need for the base
+-- Notification Handlers.
+restartCabalShakeSession :: IdeState -> VFS.VFS -> NormalizedFilePath -> String -> IO ()
```

Comment:
Nothing here seems to be cabal-specific, is this not a generic function we use elsewhere? Also why do we need it?